tunnel/intra/direct_split.go

```diff
@@ -0,0 +1,69 @@
+package intra
+
+import (
+	"io"
+	"net"
+)
+
+type splitter struct {
+	DuplexConn
+	conn  *net.TCPConn
+	fresh bool
+}
+
+// DialWithSplit returns a TCP connection that always splits the initial upstream segment.
+// Like net.Conn, it is intended for two-threaded use, with one thread calling
+// Read and CloseRead, and another calling Write, ReadFrom, and CloseWrite.
+func DialWithSplit(network string, addr *net.TCPAddr) (DuplexConn, error) {
+	conn, err := net.DialTCP(network, nil, addr)
+	if err != nil {
+		return nil, err
+	}
+
+	r := &retrier{
+		conn: conn,
+	}
+
+	return r, nil
+}
+
+// Read-related functions.
+func (s *splitter) Read(buf []byte) (int, error) {
+	return s.conn.Read(buf)
+}
+
+func (s *splitter) CloseRead() error {
+	return s.conn.CloseRead()
+}
+
+// Write-related functions
+func (s *splitter) Write(b []byte) (int, error) {
+	if s.fresh {
```

Comment:
I think flipping the condition for an early return would make this more readable.

Reply:
Done